### PR TITLE
Fix Rocksdb thread safety using refcount(no lock)

### PR DESCRIPTION
**Pull request by jiacheliu3**
### What changes are proposed in this pull request?

This change intends to fix RocksDB segfaults caused by race conditions when we close the RocksDB concurrent to ongoing operations like:

1. Iterations represented by BlockMaster.validateBlocks()
2. Backup/checkpoint where we iterate the RocksDB contents and dump into an output stream
3. RPC serving in RocksInodeStore

The above race conditions will cause the Alluxio master to segfault and crash with an hs_err.pid file like
```
Current thread (0x00007ec746db8f70):  JavaThread "Master Block Integrity Check" daemon [_thread_in_native, id=32073, stack(0x00007ebf18fd0000,0x00007ebf190d1000)]

Stack: [0x00007ebf18fd0000,0x00007ebf190d1000],  sp=0x00007ebf190cf460,  free space=1021k
Native frames: (J=compiled Java code, A=aot compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [librocksdbjni249315869067059420.so+0x5e1c4b]  rocksdb::MergingIterator::Next()+0x24b
C  [librocksdbjni249315869067059420.so+0x5ddb13]  rocksdb::MergingIterator::NextAndGetResult(rocksdb::IterateResult*)+0x13
C  [librocksdbjni249315869067059420.so+0x3cbb74]  rocksdb::DBIter::Next()+0x284
J 18876  org.rocksdb.RocksIterator.next0(J)V (0 bytes) @ 0x00007f1531d7eb73 [0x00007f1531d7eac0+0x00000000000000b3]
J 23512 c2 alluxio.master.block.DefaultBlockMaster.validateBlocks(Ljava/util/function/Function;Z)V (320 bytes) @ 0x00007f1531021cb0 [0x00007f1531021540+0x0000000000000770]
j  alluxio.master.file.DefaultFileSystemMaster.validateInodeBlocks(Z)V+11
j  alluxio.master.file.BlockIntegrityChecker.heartbeat()V+8
j  alluxio.heartbeat.HeartbeatThread.run()V+78
```

The solution is:

1. Use a flag on RocksStore to indicate that the RocksDB will be closed/restart/wiped out, so all ongoing operations will abort voluntarily. This makes sure R/W operations do not block the closer, as much as possible.
2. Use reference counting on the RocksStore. A R/W operation will increment the ref count by one, and the closer must wait for the ref count to decrement to zero (or timeout). This makes sure the closer waits for R/W to abort or complete. So there will be no possibilty of crashing.

### Why are the changes needed?

As explained above

### Does this PR introduce any user facing changes?

The change should be mostly transparent to users. However, in some small corners the behavior can be different. Before this change, if the RocksDB is closed, some list operations may return partial results or crash. After this change, the operation will fail explicitly instead.

### Performance impact

Impact on performance is further explained in https://github.com/Alluxio/alluxio/pull/17171#issuecomment-1494235026
In short, by removing the `synchronized` on `RocksStore` and switch to a thread safety model based on ref count, we should have better performance, especially when the concurrency is very high (in a large deployment with 32+ cores and 500 RPC threads).

### CODE REVIEW

File: core/common/src/main/java/alluxio/retry/CountingSleepRetry.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+/**
+ * An option which allows retrying with sleeping in between the attempts.
+ * The difference between {@code CountingSleepRetry(5, 1000)} and {@code TimeoutRetry(5000, 1000)}
+ * is that CountingSleepRetry will ignore the wall clock and JVM pause time, and wait for 5s
+ * in user time, but TimeoutRetry only waits 5s in wall clock time.
+ *
+ * So if you want to wait for
+ */
+@NotThreadSafe
+public class CountingSleepRetry implements RetryPolicy {
```
Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/retry/CountingSleepRetry.java:28]: The existing timed sleep retry implementations are all based on elapsed time (wall clock time), so in the event of long GCs, the retry doesn't really leave time for the r/w operations to proceed to abort/complete. So I figure we need a retry implementation that guarantees waiting.

File: core/common/src/main/java/alluxio/master/metastore/rocks/RocksExclusiveLockHandle.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  public RocksExclusiveLockHandle(
+      boolean isRestart, AtomicBoolean stopServingFlag, LongAdder refCount) {
+    mIsRestart = isRestart;
+    mStopServingFlag = stopServingFlag;
+    mRefCount = refCount;
+  }
+
+  @Override
+  public void close() {
+    Preconditions.checkState(mRefCount.sum() == 0,
```
Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/master/metastore/rocks/RocksExclusiveLockHandle.java:38]: I feel this check may be a bit too strict. A bug (an unfound RocksDB access which does not respect the flag) will likely crash the master process here, whereas without this check it's only a low chance to segfault.

I think the use of this check is, like the leak detector, to detect all illegal accesses in test time (especially UT). So it's ideally a property which we enable in tests, and disable in production (reduce to only log a warning).

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockMetaStore.java [hunk 3]
```diff
@@ -89,11 +95,14 @@ public RocksBlockMetaStore(String baseDir) {
     RocksDB.loadLibrary();
     // the rocksDB objects must be initialized after RocksDB.loadLibrary() is called
     mDisableWAL = new WriteOptions().setDisableWAL(true);
+    mToClose.add(mDisableWAL);
```
Comment (jiacheliu3) [inline core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockMetaStore.java:98]: now i'm adding all RocksObjects into this list instead of closing them one by one in `close()`

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksInodeStore.java [hunk 4]
```diff
@@ -108,10 +116,13 @@ public RocksInodeStore(String baseDir) {
     RocksDB.loadLibrary();
     // the rocksDB objects must be initialized after RocksDB.loadLibrary() is called
     mDisableWAL = new WriteOptions().setDisableWAL(true);
+    mToClose.add(mDisableWAL);
```
Comment (jiacheliu3) [inline core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksInodeStore.java:119]: similar to `RocksBlockMetaStore`, using this list to keep all RocksObject

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksStore.java [hunk 5]
```diff
@@ -47,41 +51,81 @@
...
+ *   }
+ * </pre></blockquote>
+ *
+ * For operations like closing/restart/restoring on the RocksDB, an exclusive lock should
+ * be acquired by calling one of:
+ * 1. {@link #lockForClosing()}
+ * 2. {@link #lockForRestart()}
  */
-@ThreadSafe
+@NotThreadSafe
```
Comment (jiacheliu3) [inline core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksStore.java:93]: Before: thread safety by `synchronized` methods and `@ThreadSafe` tag, however the thread safety beyond the `synchronized` call is flawed
After: removed embedded thread safety and removed `synchronized`. The thread safety is provided by providing locking methods, and the caller should invoke those calls in the correct way.

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksInodeStore.java [hunk 6]
```diff
@@ -514,20 +566,24 @@ public CheckpointName getCheckpointName() {
 
   @Override
   public void writeToCheckpoint(OutputStream output) throws IOException, InterruptedException {
-    mRocksStore.writeToCheckpoint(output);
+    try (RocksExclusiveLockHandle lock = mRocksStore.lockForRestart()) {
```
Comment (jiacheliu3) [inline core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksInodeStore.java:569]: Writing to checkpoint will acquire an exclusive lock and therefore all ongoing operations will be aborted. It does sound a bit awkward if we phrase it this way.

However, I think this does not really happen, where "this" means taking a checkpoint will fail all ongoing iterations. First of all, the shared state lock will be acquired before taking the checkpoint. And by being able to take that lock, we've already made sure there's no RPC. Also, on standby masters where checkpoints are taken, there will not be ongoing RPCs.

**Comment by jiacheliu3** [comment]
Below are some JMH micro benchmarks on different lock implementations. The benchmark was done on my Mac with 2.6GHz 6-core i7 and 16G RAM, but it was also running many other applications so it's nowhere close to a server.
```
Benchmark                    Mode  Cnt      Score      Error  Units
RocksStoreBench.getReadLock  avgt   25  74210.940 ± 3391.482  us/op # Use ReentrantReadWriteLock
RocksStoreBench.getReferenceLock  avgt   25  19981.210 ± 875.339  us/op # Using AtomicInteger as ref count
RocksStoreBench.getSynchronized  avgt   25  20499.528 ± 940.156  us/op # Use synchronized on RocksStore(not really thread safe)
RocksStoreBench.getReferenceLock  avgt   25  828.177 ± 6.636  us/op # Using LongAdder as ref count
```

So our final decision was to use the LongAdder for ref count, and rely on the ref count with no lock, for optimized performance.

File: core/server/master/src/test/java/alluxio/master/metastore/rocks/RocksInodeStoreTest.java [hunk 7]
```diff
@@ -14,42 +14,394 @@
...
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ThreadLocalRandom;
 
 public class RocksInodeStoreTest {
   @Rule
   public TemporaryFolder mFolder = new TemporaryFolder();
 
+  public RocksInodeStore mStore;
```
Comment (jiacheliu3) [inline core/server/master/src/test/java/alluxio/master/metastore/rocks/RocksInodeStoreTest.java:53]: I added some UT on how the ref count behaves correctly under some concurrency. @tcrain could you leave some comments on the test cases if you have some on top of your head? Also pls ignore how messy my code currently is. I intend to refactor code in this file once we settle on the test cases. Thx!

Right now all test cases here are passing.

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksInodeStore.java (hunk 6 above)
Comment (tcrain) [reply core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksInodeStore.java:572]: Yes there should be no operations happening during checkpointing or restoring from checkpointing, otherwise the checkpoint would be invalid anyway

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksStore.java [hunk 8]
```diff
@@ -347,4 +398,131 @@ private static IndexType toRocksIndexType(
...
+    * 7. Reader decrements refCount aborts in checkAndAcquireSharedLock()
+    */
+    Instant waitStart = Instant.now();
+    CountingRetry retry = new CountingRetry((int) ROCKS_CLOSE_WAIT_TIMEOUT.getSeconds() * 10);
+    while (mRefCount.sum() != 0 && retry.attempt()) {
+      SleepUtils.sleepMs(100);
+    }
+    Duration elapsed = Duration.between(waitStart, Instant.now());
+    LOG.info("Waited {}ms for ongoing read/write to complete/abort", elapsed.toMillis());
+    System.out.println("Waited " + elapsed.toMillis() + " for ongoing read/write to complete/abort");
```
Comment (tcrain) [inline core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksStore.java:467]: Should remove this line?

**Comment by tcrain** [comment]
My only worry is when a BlockStore or InodeStore is reset.

If there is a really slow thread, when it tries for example to get the next item in its iterator, the store may have closed and reset in the meantime, so that it thinks it can run successfully.

For this I would say instead of reseting underlying RocksStore object, just close the old one and create a new one.

File: core/common/src/main/java/alluxio/master/metastore/rocks/RocksExclusiveLockHandle.java (hunk 2 above)
Comment (jiacheliu3) [reply core/common/src/main/java/alluxio/master/metastore/rocks/RocksExclusiveLockHandle.java:38]: In prod environment we want to be tolerant and lenient to small data races. In other words if the sum is not zeroed after a timeout, we print a warning and proceed (instead of crashing).

But in tests we want to be very strict, meaning if there's a race we want the process to crash in a super obvious way so we find the issue and fix it. 

That difference suggests we want have a config, like `alluxio.leak.detector.level`. We use `alluxio.leak.detector.level=DISABLED` in real deployment but use `alluxio.leak.detector.level=PARANIOD` in tests. `TEST_MODE` happens to be such a config i think I'm going to use here.

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockMetaStore.java (hunk 3 above)
Comment (jiacheliu3) [reply core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockMetaStore.java:98]: Actually i prefer this style where we create something and immediately put it into the record. Imagine my brain only has a register of size one.  And I will panic if i don't deal with this obj immediately because i know I for sure will forget.

File: core/common/src/main/java/alluxio/master/metastore/rocks/RocksUtils.java [hunk 9]
```diff
@@ -111,23 +129,38 @@ public boolean hasNext() {
...
             valid.set(false);
+            rocksIterator.close();
           }
         }
       }
     };
 
-    return CloseableIterator.create(iter, (whatever) -> rocksIterator.close());
+    return CloseableIterator.create(iter, (whatever) -> {
+      rocksIterator.close();
```
[the remark this answers is not in the document]
Comment (jiacheliu3) [reply core/common/src/main/java/alluxio/master/metastore/rocks/RocksUtils.java:159]: That is a brilliant question!

According to https://stackoverflow.com/a/15895573/4933827
The `close()` statement is handled in the `catch/finally` block of the caller. That means I need to guard the callers of this CloseableIterator and catch there.

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksInodeStore.java [hunk 10]
```diff
@@ -353,37 +370,51 @@ public Optional<MutableInode<?>> getMutable(long id, ReadOption option) {
...
-      iter.seek(RocksUtils.toByteArray(inodeId, seekTo));
+      if (seekTo != null && seekTo.length() > 0) {
+        iter.seek(RocksUtils.toByteArray(inodeId, seekTo));
+      }
+      /*
+       * Acquire a second lock for iteration, instead of using the same lock for initialization.
+       * Because init takes many operations and should be protected by try-with-resource.
+       * This is fine because the shared lock is reentrant.
+       */
+      RocksSharedLockHandle readLock = mRocksStore.checkAndAcquireSharedLock();
```
[the remark this answers is not in the document]
Comment (jiacheliu3) [reply core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksInodeStore.java:400]: yes, 1st increment is for creating the iterator, 2nd increment is creating the lock for the iterator to use. These two "locks"(increments) hold hands to there's no gap in the middle and everything are guarded. 1st "lock" is to protect the RocksObject instances used when creating the iter.

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksStore.java (hunk 8 above)
Comment (jiacheliu3) [reply core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksStore.java:467]: eagle eye

File: core/common/src/main/java/alluxio/master/metastore/rocks/RocksUtils.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (jiacheliu3) [reply core/common/src/main/java/alluxio/master/metastore/rocks/RocksUtils.java:159]: If the question is if this `rocksIterator.close();` throws an exception then `rocksDbSharedLock.close();` at L162 will not execute, that is indeed a concern. 

There are some reasons why the code looks like this:
1. `rocksIterator.close()` throws no checked exception, theoretically it only throws RuntimeException and I don't know what they can be.
2. This lambda is executed in a try-with-resource block at the caller, having a nested try-finally block here is just awkward. 
3. Moving L162 above L159 is incorrect.

**Comment by jiacheliu3** [comment]
> 

You mean something like the following sequence right?
```
1. Reader gets the read lock (refCount++) then go to sleep
2. Closer sees the refCount=1 then wait
3. Closer wait times out and forces the lock (the refCount reset to zero)
4. Sometime later and something has happened, the reader wakes up now and there are multiple problems!
```

Problem 1: Obsolete references

No matter what you do (like replacing the `RocksStore` object in `RocksBlockMetaStore`), there's still a chance that sleeping reader holds an old `RocksDB` object from `RocksStore.getDb()` or an old `RocksIterator` created over old column object etc. I think you can replace all member variables from `RocksStore` but cannot replace the references that reader thread is holding, and there will always be a check-then-access race?

I think whenever the exclusive lock is forced (i.e. ignoring a non-zero refcount), there's risk for such slow readers to not check and crash the process with an obsolete reference (the check happened before the sleep). The time-out we added to the exclusive lock simply means we give up on this situation. In terms of the trade-off, i like the timeout idea and i think we can take the risk and fix issues when they arise.

Problem 2: ABA problem

The reader can wake up and see the RocksDB running, not knowing about the reset at all. I think I can fix this by adding a version somewhere, if that's necessary.

Problem 3: Ref count

When the exclusive lock forces locking, it MUST reset the ref count for fear that the incorrect count is because someone forgot to decrement. But if there's really a super slow reader, the decrement will ruin the ref count again (well incorrect ref count is honestly not end of the world anyways). I think a similar versioning is also needed, where the decrement need to double check the version.

*Conclusion* 

I don't want to find excuses and say these are all acceptible or there's nothing we can do, but I do feel that we are introducing a somewhat complex mechanism and it will just be much more complex if we handle all the corner cases. So my gut feeling is i will stop at a point where the performance goes bad or it's too hard to understand. And the risk will be compensated elsewhere like:
1. Where is the slow reader from, can we get rid of that?
2. From the system level, can we for example explicitly stop those heartbeat thread cycles when we restore metadata. Identify dangling threads and slow readers and fix them.
3. Are there other locks that helps us make sure the exclusive lock is never forced?

**Comment by jiacheliu3** [comment]
> My only worry is when a BlockStore or InodeStore is reset.
> 
> If there is a really slow thread, when it tries for example to get the next item in its iterator, the store may have closed and reset in the meantime, so that it thinks it can run successfully.
> 
> For this I would say instead of reseting underlying RocksStore object, just close the old one and create a new one.

In terms of creating a new ref, is this `RocksStore.createDb()` doing what you want? https://github.com/Alluxio/alluxio/blob/e4499aeadad24f6e548eda017d18aeb33c7d9de7/core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksStore.java#L179

File: core/common/src/main/java/alluxio/master/metastore/rocks/RocksUtils.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (jiacheliu3) [reply core/common/src/main/java/alluxio/master/metastore/rocks/RocksUtils.java]: Actually i found that a try-finally block in java is cheap, so i updated the logic to release the lock in finally

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksInodeStore.java [hunk 11]
```diff
@@ -327,7 +344,7 @@ public void addChild(long parentId, String childName, Long childId) {
 
   @Override
   public void removeChild(long parentId, String name) {
-    try {
+    try (RocksSharedLockHandle lock = mRocksStore.checkAndAcquireSharedLock()){
       db().delete(mEdgesColumn.get(), mDisableWAL, RocksUtils.toByteArray(parentId, name));
```
Comment (tcrain) [inline core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksInodeStore.java:348]: I'm wondering if we can move all the handling of the concurrency handling to inside the RocksStore object? For example here the lock check could take place inside the db().delete() call?
Maybe also then the mEdgesColumn lifecycle could be handled by the RocksStore as well?

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksInodeStore.java [hunk 12]
```diff
@@ -493,14 +538,29 @@ public Set<MutableInode<?>> allInodes() {
...
-        db().newIterator(mInodesColumn.get(), mIteratorOption),
-        (iter) -> getMutable(Longs.fromByteArray(iter.key()), ReadOption.defaults()).get());
+    try (RocksSharedLockHandle lock = mRocksStore.checkAndAcquireSharedLock()) {
+      RocksSharedLockHandle readLock = mRocksStore.checkAndAcquireSharedLock();
+      return RocksUtils.createCloseableIterator(
+          db().newIterator(mInodesColumn.get(), mIteratorOption),
+          (iter) -> getMutable(Longs.fromByteArray(iter.key()), ReadOption.defaults()).get(),
+          // TODO(jiacheng): validate an ex is thrown in UT
+          () -> {
+            mRocksStore.shouldAbort(lock.mDbVersion);
```
Comment (tcrain) [inline core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksInodeStore.java:560]: Or here, could it be possible to move the abort checks to inside the next() and hasNext() calls of the iterator created by the call db().newIterator() in RocksStore?

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksInodeStore.java (hunk 11 above)
Comment (jiacheliu3) [reply core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksInodeStore.java:348]: The `delete()` is in `RocksDB` class, which is in the RocksJava lib we don't have control over. So we can't easily move all these tailored logic into RocksStore trivially. However, there are ways we can achieve that, and instead of addressing in this PR, I will backlog this and address that as a design debt in 3.0/3.x. https://github.com/Alluxio/alluxio/issues/17271

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksInodeStore.java (hunk 12 above)
Comment (jiacheliu3) [reply core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksInodeStore.java:560]: Same as https://github.com/Alluxio/alluxio/pull/17171#discussion_r1169062360 we don't have control over `db().newIterator()` and we can only wrap around that.

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksStore.java [hunk 13]
```diff
@@ -374,4 +465,236 @@ private static IndexType toRocksIndexType(
...
+        throw new RuntimeException(ExceptionMessage.ROCKS_DB_EXCLUSIVE_LOCK_FORCED
+            .getMessage(unclosedOperations));
+      }
+      /*
+       * Set the flag so shared locks know that the ref count has been reset,
+       * no need to update the ref count on unlock.
+       * If one shared lock did not decrement the ref count before this reset, it should not
+       * decrement the ref count when it is released.
+       */
+      mRefCount.reset();
```
[the remark this answers is not in the document]
Comment (jiacheliu3) [reply core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksStore.java:615]: makes sense to me, it's much cleaner that way i'll try to implement it

**Review by tcrain — APPROVED** [review]

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please

**Comment by yuzhu** [comment]
alluxio-bot, cherry-pick this to branch-2.8 please
